File: wisefy/src/androidTest/java/com/isupatches/wisefy/DisconnectFromCurrentNetworkTests.java
package com.isupatches.wisefy;


import com.isupatches.wisefy.callbacks.DisconnectFromCurrentNetworkCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import org.junit.Test;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import static org.mockito.Mockito.after;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class DisconnectFromCurrentNetworkTests extends BaseAndroidJUnit4TestClass {

    @Test
    public void sync_failure() {
        when(mMockWiFiManager.disconnect()).thenReturn(false);
        assertEquals(false, mWiseFy.disconnectFromCurrentNetwork());
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnect();
    }

    @Test
    public void sync_failure_missingPrerequisite() {
        missingPrerequisite();
        assertEquals(false, mWiseFy.disconnectFromCurrentNetwork());
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).disconnect();
    }

    @Test
    public void sync_success() {
        when(mMockWiFiManager.disconnect()).thenReturn(true);
        assertEquals(true, mWiseFy.disconnectFromCurrentNetwork());
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnect();
    }

    @Test
    public void async_failure() {
        when(mMockWiFiManager.disconnect()).thenReturn(false);
        DisconnectFromCurrentNetworkCallbacks mockCallbacks = mock(DisconnectFromCurrentNetworkCallbacks.class);
        mWiseFy.disconnectFromCurrentNetwork(mockCallbacks);
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnect();
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureDisconnectingFromCurrentNetwork();
    }

    @Test
    public void async_failure_nullCallback() {
        when(mMockWiFiManager.disconnect()).thenReturn(false);
        try {
            mWiseFy.disconnectFromCurrentNetwork(null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnect();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_missingPrerequisite() {
        missingPrerequisite();
        DisconnectFromCurrentNetworkCallbacks mockCallbacks = mock(DisconnectFromCurrentNetworkCallbacks.class);
        mWiseFy.disconnectFromCurrentNetwork(mockCallbacks);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).disconnect();
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnectFromCurrentNetworkWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
    }

    @Test
    public void async_failure_missingPrerequisite_nullCallback() {
        missingPrerequisite();
        try {
            mWiseFy.disconnectFromCurrentNetwork(null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).disconnect();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_success() {
        when(mMockWiFiManager.disconnect()).thenReturn(true);
        DisconnectFromCurrentNetworkCallbacks mockCallbacks = mock(DisconnectFromCurrentNetworkCallbacks.class);
        mWiseFy.disconnectFromCurrentNetwork(mockCallbacks);
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnect();
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnectedFromCurrentNetwork();
    }

    @Test
    public void async_success_nullCallback() {
        when(mMockWiFiManager.disconnect()).thenReturn(true);
        try {
            mWiseFy.disconnectFromCurrentNetwork(null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnect();
        } catch (NullPointerException npe) {
            fail();
        }
    }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/DisconnectFromCurrentNetworkTests.kt
package com.isupatches.wisefy.main

import com.isupatches.wisefy.BaseAndroidJUnit4TestClass
import com.isupatches.wisefy.VERIFICATION_SUCCESS_TIMEOUT
import com.isupatches.wisefy.callbacks.DisconnectFromCurrentNetworkCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER

import org.junit.Assert.assertEquals
import org.junit.Test

import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * Tests the ability to disconnect a device from it's current network.
 *
 * @author Patches
 */
internal class DisconnectFromCurrentNetworkTests : BaseAndroidJUnit4TestClass() {

    @Test fun sync_failure_prechecks() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_failure()
        assertEquals(false, wisefy.disconnectFromCurrentNetwork())
        verificationUtil.didNotTryToDisconnectFromCurrentNetwork()
    }

    @Test fun sync_failure() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_success()
        mockNetworkUtil.disconnectFromCurrentNetwork_failure()
        assertEquals(false, wisefy.disconnectFromCurrentNetwork())
        verificationUtil.triedToDisconnectFromCurrentNetwork()
    }

    @Test fun sync_success() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_success()
        mockNetworkUtil.disconnectFromCurrentNetwork_success()
        assertEquals(true, wisefy.disconnectFromCurrentNetwork())
        verificationUtil.triedToDisconnectFromCurrentNetwork()
    }

    @Test fun async_failure_prechecks() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_failure()
        val mockCallbacks = mock(DisconnectFromCurrentNetworkCallbacks::class.java)
        wisefy.disconnectFromCurrentNetwork(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
        verificationUtil.didNotTryToDisconnectFromCurrentNetwork()
    }

    @Test fun async_failure_prechecks_nullCallback() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_failure()
        nullCallbackUtil.callDisconnectFromCurrentNetwork()
        verificationUtil.didNotTryToDisconnectFromCurrentNetwork()
    }

    @Test fun async_failure() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_success()
        mockNetworkUtil.disconnectFromCurrentNetwork_failure()
        val mockCallbacks = mock(DisconnectFromCurrentNetworkCallbacks::class.java)
        wisefy.disconnectFromCurrentNetwork(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureDisconnectingFromCurrentNetwork()
        verificationUtil.triedToDisconnectFromCurrentNetwork()
    }

    @Test fun async_failure_nullCallback() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_success()
        mockNetworkUtil.disconnectFromCurrentNetwork_failure()
        nullCallbackUtil.callDisconnectFromCurrentNetwork()
        verificationUtil.triedToDisconnectFromCurrentNetwork()
    }

    @Test fun async_success() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_success()
        mockNetworkUtil.disconnectFromCurrentNetwork_success()
        val mockCallbacks = mock(DisconnectFromCurrentNetworkCallbacks::class.java)
        wisefy.disconnectFromCurrentNetwork(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).disconnectedFromCurrentNetwork()
        verificationUtil.triedToDisconnectFromCurrentNetwork()
    }

    @Test fun async_success_nullCallback() {
        mockWiseFyPrechecksUtil.disconnectFromCurrentNetwork_success()
        mockNetworkUtil.disconnectFromCurrentNetwork_success()
        nullCallbackUtil.callDisconnectFromCurrentNetwork()
        verificationUtil.triedToDisconnectFromCurrentNetwork()
    }
}
